pkg/chunk/chunk.go

```diff
@@ -253,7 +271,18 @@ func (c *Chunk) Decode(input []byte) error {
 			return errors.WithStack(ErrWrongMetadata)
 		}
 	}
-	*c = tempMetadata
+	*c = tempMetadata.Chunk
+
+	// If we have decoded a chunk with the same fingerprint before, re-use its Metric, otherwise parse it
+	if cachedMetric, found := decodeContext.metrics[c.Fingerprint]; found {
+		c.Metric = cachedMetric
+	} else {
+		err := json.NewDecoder(bytes.NewReader(tempMetadata.RawMetric)).Decode(&c.Metric)
+		if err != nil {
+			return errors.Wrap(err, "while parsing chunk metric")
+		}
+		decodeContext.metrics[c.Fingerprint] = c.Metric
+	}
```

Comment:
I'd personally make this bit a method on `DecodeContext`, but no need if you don't want to.